pkg/sql/catalog/tabledesc/constraint.go

```diff
@@ -11,215 +11,477 @@
+func (c foreignKeyConstraint) IsConstraintUnvalidated() bool {
+	return c.desc.Validity == descpb.ConstraintValidity_Unvalidated
+}
+
+// GetName implements the catalog.Constraint interface.
+func (c foreignKeyConstraint) GetName() string {
+	return c.desc.Name
+}
+
+// AsForeignKey implements the catalog.ConstraintProvider interface.
+func (c foreignKeyConstraint) AsForeignKey() catalog.ForeignKeyConstraint {
+	return &c
+}
+
+// String implements the catalog.Constraint interface.
+func (c foreignKeyConstraint) String() string {
+	return fmt.Sprintf("%+v", c.desc)
+}
+
+// IsEnforced implements the catalog.Constraint interface.
+func (c foreignKeyConstraint) IsEnforced() bool {
+	return !c.IsMutation() || c.WriteAndDeleteOnly()
+}
+
+// constraintCache contains precomputed slices of constraints, categorized by:
+//   - constraint subtype: checks, fks, etc.
+//   - enforcement status: whether a constraint is enforced for data written to
+//     the table, regardless of whether the constraint is valid for table data
+//     which existed prior to the constraint being added to the table.
+type constraintCache struct {
+	all, allEnforced       []catalog.Constraint
+	checks, checksEnforced []catalog.CheckConstraint
+	fks, fksEnforced       []catalog.ForeignKeyConstraint
+	uwis, uwisEnforced     []catalog.UniqueWithIndexConstraint
+	uwois, uwoisEnforced   []catalog.UniqueWithoutIndexConstraint
+	fkBackRefs             []catalog.ForeignKeyConstraint
+}
+
+// newConstraintCache returns a fresh fully-populated constraintCache struct for the
+// TableDescriptor.
+func newConstraintCache(
+	desc *descpb.TableDescriptor, indexes *indexCache, mutations *mutationCache,
+) *constraintCache {
+	capUWIs := len(indexes.all)
+	numEnforcedChecks := len(desc.Checks)
+	capChecks := numEnforcedChecks + len(mutations.checks)
+	numEnforcedFKs := len(desc.OutboundFKs)
+	capFKs := numEnforcedFKs + len(mutations.fks)
+	numEnforcedUWOIs := len(desc.UniqueWithoutIndexConstraints)
+	capUWOIs := numEnforcedUWOIs + len(mutations.uniqueWithoutIndexes)
+	capAll := capChecks + capFKs + capUWOIs + capUWIs
+	// Pre-allocate slices which are known not to be empty:
+	// physical tables always have at least one index-backed unique constraint
+	// in the form of the primary key.
+	c := constraintCache{
+		all:          make([]catalog.Constraint, 0, capAll),
+		allEnforced:  make([]catalog.Constraint, 0, capAll),
+		uwis:         make([]catalog.UniqueWithIndexConstraint, 0, capUWIs),
+		uwisEnforced: make([]catalog.UniqueWithIndexConstraint, 0, capUWIs),
 	}
-	for _, cst := range append(desc.OutboundFKs, desc.InboundFKs...) {
-		backingStruct := constraint{
-			desc: &descpb.ConstraintToUpdate{
-				ConstraintType: descpb.ConstraintToUpdate_FOREIGN_KEY,
-				Name:           cst.Name,
-				ForeignKey:     cst,
-			},
+	// Populate with index-backed unique constraints.
+	for _, idx := range indexes.all {
+		if uwi := idx.AsUniqueWithIndex(); uwi != nil && uwi.NumKeyColumns() > 0 {
```

Comment:
why do we check `uwi.NumKeyColumns()` additionally? In what scenario will we have a non-nil uwi but 0 key columns?

Reply:
Virtual tables. It's dumb, in those cases we should have `GetPrimaryIndex()` return `nil`, yet we don't. Yet more tech debt.